### PR TITLE
util: increase several macros limit from 255 to 4095

### LINKED ISSUE #63497: arch: number of interrupt source is limited to 255
**Describe the bug**
[RISCV PLIC specification](https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc#3-interrupt-priorities) states that the controller can support 1024 interrupt sources:

> Total 1024 Interrupt Source Priority in this memory block.

However the `LISTIFY` macro used in the `arch/common/isr_tables.c` only support length of up to 255:

https://github.com/zephyrproject-rtos/zephyr/blob/c461441cc5aac2738de17ff89e46214eb2ae5750/arch/common/isr_tables.c#L63-L65

A board that has more than 255 IRQNs (as defined by `CONFIG_NUM_IRQS`) can't be compiled, see the build log below.

**Please also mention any information which could help others to understand
the problem you're facing**
 - Any RISCV board with PLIC and support more than 255 IRQs

**To Reproduce**
1. Modify the IRQ number devicetree & defconfig of `qemu_riscv32` board to 1024
2. `west build -b qemu_riscv32 -p auto -t run -T zephyr/tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_direct`
3. See build error

**Expected behavior**
Any RISCV board with PLIC should be able to support up the number of interrupt specified in the vendor datasheet & according to PLIC specifications.

**Impact**
showstopper for RISCV board with PLIC and requires more than 255 IRQs.

**Logs and console output**
```
In file included from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/atomic.h:16,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/kernel_structs.h:24,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/arch/riscv/arch_inlines.h:12,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/arch/arch_inlines.h:26,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/arch_interface.h:1225,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/arch/cpu.h:12,
                 from /Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c:10:
/Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c: In function '_irq_vector_table':
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:418:44: warning: implicit declaration of function 'Z_UTIL_LISTIFY_1024'; did you mean 'Z_UTIL_LISTIFY_104'? [-Wimplicit-function-declaration]
  418 | #define LISTIFY(LEN, F, sep, ...) UTIL_CAT(Z_UTIL_LISTIFY_, LEN)(F, sep, __VA_ARGS__)
      |                                            ^~~~~~~~~~~~~~~
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in definition of macro 'UTIL_PRIMITIVE_CAT'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:418:35: note: in expansion of macro 'UTIL_CAT'
  418 | #define LISTIFY(LEN, F, sep, ...) UTIL_CAT(Z_UTIL_LISTIFY_, LEN)(F, sep, __VA_ARGS__)
      |                                   ^~~~~~~~
/Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c:64:10: note: in expansion of macro 'LISTIFY'
   64 |          LISTIFY(CONFIG_NUM_IRQS, BUILD_VECTOR, (;));
      |          ^~~~~~~
/Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c:64:35: error: 'BUILD_VECTOR' undeclared (first use in this function)
   64 |          LISTIFY(CONFIG_NUM_IRQS, BUILD_VECTOR, (;));
      |                                   ^~~~~~~~~~~~
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:418:66: note: in definition of macro 'LISTIFY'
  418 | #define LISTIFY(LEN, F, sep, ...) UTIL_CAT(Z_UTIL_LISTIFY_, LEN)(F, sep, __VA_ARGS__)
      |                                                                  ^
/Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c:64:35: note: each undeclared identifier is reported only once for each function it appears in
   64 |          LISTIFY(CONFIG_NUM_IRQS, BUILD_VECTOR, (;));
      |                                   ^~~~~~~~~~~~
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:418:66: note: in definition of macro 'LISTIFY'
  418 | #define LISTIFY(LEN, F, sep, ...) UTIL_CAT(Z_UTIL_LISTIFY_, LEN)(F, sep, __VA_ARGS__)
      |                                                                  ^
/Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c:64:50: error: expected expression before ';' token
   64 |          LISTIFY(CONFIG_NUM_IRQS, BUILD_VECTOR, (;));
      |                                                  ^
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:418:69: note: in definition of macro 'LISTIFY'
  418 | #define LISTIFY(LEN, F, sep, ...) UTIL_CAT(Z_UTIL_LISTIFY_, LEN)(F, sep, __VA_ARGS__)
      |                                                                     ^~~
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:418:85: error: expected expression before ')' token
  418 | #define LISTIFY(LEN, F, sep, ...) UTIL_CAT(Z_UTIL_LISTIFY_, LEN)(F, sep, __VA_ARGS__)
      |                                                                                     ^
/Users/ycsin/zephyrproject/zephyr/arch/common/isr_tables.c:64:10: note: in expansion of macro 'LISTIFY'
   64 |          LISTIFY(CONFIG_NUM_IRQS, BUILD_VECTOR, (;));
      |          ^~~~~~~
[44/106] Linking C static library zephyr/libzephyr.a
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /Users/ycsin/homebrew/bin/cmake --build /Users/ycsin/zephyrproject/build --target run
```

**Environment (please complete the following information):**
 - OS: MacOS
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: c461441cc5aac2738de17ff89e46214eb2ae5750

### PR DESCRIPTION
**Pull request by ycsin**
Currently, the following macros will only work for 0 - 255:
- `IS_EQ`
- `UTIL_X2`
- `UTIL_INC`
- `UTIL_DEC`
- `LISTIFY`

This patch increases their limit to 4095.

fixes #63497

### CODE REVIEW

File: include/zephyr/sys/util_internal_util_inc.h [hunk 1]
```diff
@@ -0,0 +1,4113 @@
...
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_H_
+#error "This header should not be used directly, please include util_internal.h instead"
+#endif /* ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_H_ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_UTIL_INC_H_
+#define ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_UTIL_INC_H_
+
+#define UTIL_INC_0 1
```
Comment (kartben) [inline include/zephyr/sys/util_internal_util_inc.h:16]: this was already an issue prior to this patch but these `UTIL_INC_*` macros should actually be marked internal -- either use `Z_` prefix like for e.g `Z_LISTIFY`, or wrap in a ` @cond INTERNAL_HIDDEN` section.

File: include/zephyr/sys/util_internal_util_dec.h [hunk 2]
```diff
@@ -0,0 +1,4113 @@
...
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_H_
+#error "This header should not be used directly, please include util_internal.h instead"
+#endif /* ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_H_ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_UTIL_DEC_H_
+#define ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_UTIL_DEC_H_
+
+#define UTIL_DEC_0 0
```
Comment (kartben) [inline include/zephyr/sys/util_internal_util_dec.h:16]: ditto re: marking internal

File: include/zephyr/sys/util_internal_util_x2.h [hunk 3]
```diff
@@ -0,0 +1,4113 @@
...
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_H_
+#error "This header should not be used directly, please include util_internal.h instead"
+#endif /* ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_H_ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_UTIL_X2_H_
+#define ZEPHYR_INCLUDE_SYS_UTIL_INTERNAL_UTIL_X2_H_
+
+#define UTIL_X2_0 0
```
Comment (kartben) [inline include/zephyr/sys/util_internal_util_x2.h:16]: ditto re: marking internal

**Review by kartben — DISMISSED** [review]

**Comment by cfriedt** [comment]
Wow - good catch 👍

**Comment by ycsin** [comment]
Adding the `Trivial` tag as this patch basically just extend existing macros from 255 to 4095 at no cost to the firmware complexity

**Review by cfriedt — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**